Beheim/superdense coding

bettinaheim opened the pull request:
I did some minor updates to the superdense coding kata related to a demo. 
In particular, I switched to using a user defined type Message to represent the two bits of classical information to transmit.

cgranade commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -3,11 +3,16 @@
+    /// Data type that represents the message that is transmitted 
+    /// as part of the superdense coding protocol. 
+    /// It includes two classical bits (Bit1 and Bit2) of type Bool.   
+    newtype Message = (Bit1 : Bool, Bit2 : Bool);    
```
Minor thought, but do we want to use the Alice and Bob naming scheme here as well?

bettinaheim replied:
For the message? That to me seems more confusing rather than helpful, since it sounds like it implies that one bit belongs to Alice, and one to Bob.

cgranade commented on SuperdenseCoding/Tests.qs:
```diff
@@ -14,57 +14,50 @@ namespace Quantum.Kata.SuperdenseCoding {
-        using (qs = Qubit[N]) {
+        
+    operation T1_CreateEntangledPair_Test () : Unit {
+        using ((q1,q2) = (Qubit(), Qubit())) {
```
```suggestion
        using ((q1, q2) = (Qubit(), Qubit())) {
```

cgranade commented on SuperdenseCoding/Tests.qs:
```diff
@@ -14,57 +14,50 @@ namespace Quantum.Kata.SuperdenseCoding {
+
             // assert that all qubits end up in |0⟩ state
-            AssertAllZero(qs);
+            AssertAllZero([q1,q2]);
```
```suggestion
            AssertAllZero([q1, q2]);
```

tcNickolas commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -3,11 +3,16 @@
+    /// Data type that represents the message that is transmitted 
+    /// as part of the superdense coding protocol. 
+    /// It includes two classical bits (Bit1 and Bit2) of type Bool.   
+    newtype Message = (Bit1 : Bool, Bit2 : Bool);    
```
I believe the compiler doesn't mind, but for me type `Message` just reads as logging function `Message` and is confusing... Could we call it `ProtocolMessage` instead?

bettinaheim replied:
Sure.

tcNickolas commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -40,22 +45,19 @@ namespace Quantum.Kata.SuperdenseCoding {
     // Task 1. Entangled pair
     // Input: An array of two qubits in the |00⟩ state.
```
Please update input description to match operation signature.

bettinaheim replied:
I updated the description. Take a look if you are happy with it.

tcNickolas commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -68,21 +70,16 @@ namespace Quantum.Kata.SuperdenseCoding {
+    // Task 3. Decode the message and reset the qubits (Bob's task)
+    // Decode the message using the qubit received from Alice and reset them to a |00⟩ state.
     // Inputs:
     //      1) Bob's part of the entangled pair qBob.
```
Please update input description (order of inputs) to match operation signature.

tcNickolas commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -68,21 +70,16 @@ namespace Quantum.Kata.SuperdenseCoding {
-    // Task 3. Decode the message (Bob's task)
-    // Decode the message using the qubit received from Alice.
+    // Task 3. Decode the message and reset the qubits (Bob's task)
+    // Decode the message using the qubit received from Alice and reset them to a |00⟩ state.
```
```suggestion
    // Decode the message using the qubit received from Alice and reset both qubits to a |00⟩ state.
```

tcNickolas commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -40,22 +45,19 @@ namespace Quantum.Kata.SuperdenseCoding {
     // Inputs:
     //      1) Alice's part of the entangled pair of qubits qAlice.
-    //      2) two classical bits, stored in an array.
+    //      2) two classical bits, stored as Message.
```
This is the first time people see a type, so I'd suggest giving a quick explanation of how they can access the bits (`message::Bit1`).

bettinaheim replied:
I added the first if-statement (without body) to the implementation stump below and commented that it accesses Bit1 of message. Take a look and let me know if that's what you had in mind, or if you have a better idea on how to explain named item access.

tcNickolas commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -68,21 +70,16 @@ namespace Quantum.Kata.SuperdenseCoding {
+    operation DecodeMessageFromQubits (qAlice : Qubit, qBob : Qubit) : Message {
 
-        return decoded_bits;
+        fail ("not implemented");
```
```suggestion
        fail ("Task 3 not implemented");
```

tcNickolas commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -93,13 +90,9 @@ namespace Quantum.Kata.SuperdenseCoding {
-        return decoded_bits;
+    operation SuperdenseCodingProtocol (message : Message) : Message {
+                
+        fail ("not implemented");
```
```suggestion
        fail ("Task 4 not implemented");
```

tcNickolas commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -10,7 +10,7 @@ namespace Quantum.Kata.SuperdenseCoding {
     /// as part of the superdense coding protocol. 
     /// It includes two classical bits (Bit1 and Bit2) of type Bool.   
-    newtype Message = (Bit1 : Bool, Bit2 : Bool);    
+    newtype ProtocollMessage = (Bit1 : Bool, Bit2 : Bool);    
```
Typo: should be `ProtocolMessage` (with one `l`)

tcNickolas commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -38,24 +43,22 @@ namespace Quantum.Kata.SuperdenseCoding {
-        // You don't need to modify them. Feel free to remove them, this won't cause your code to fail.
-        EqualityFactI(Length(qs), 2, "The array should have exactly 2 qubits.");
+    // Input: A tuple of two qubits.
+    // Goal:  If both qubits are initially in a state |00>, 
```
This task description is incorrect/incomplete; we don't ask people to figure out whether the qubits are in state 00 and modify the state only if they are, we ask them to prepare a certain state and guarantee that the qubits start in 00.

If you want to modify this task to ask for a unitary transformation H-CNOT, you need to properly describe it, and to modify the test `T1_CreateEntangledPair_Test` to use `AssertOperationsEqualReferenced` instead of `AssertEqualOnZeroState` to make sure that it checks exactly the same thing as the task asks to do.

If you want to keep this task as is (which, to be fair, is what I would do - we don't *force* people to use adjoint of this task to perform later stages of the teleportation protocol, and formulating the task as state preparation doesn't prevent *us* from doing it our way and using it later), keep the goal as it was and modify input to "Two qubits, each in the |0⟩ state."

(Also, note the use of fancy Unicode |0⟩ instead of plain ASCII |0> to denote states :-) )

bettinaheim replied:
Fair point. As suggested, I kept the original task goal and modified the input description to "Two qubits, each in the |0⟩ state.".

tcNickolas commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -38,24 +43,22 @@ namespace Quantum.Kata.SuperdenseCoding {
-        // The following lines enforce the constraints on the input that you are given.
-        // You don't need to modify them. Feel free to remove them, this won't cause your code to fail.
-        EqualityFactI(Length(qs), 2, "The array should have exactly 2 qubits.");
+    // Input: A tuple of two qubits.
```
```suggestion
    // Input: Two qubits.
```

tcNickolas commented on SuperdenseCoding/Tasks.qs:
```diff
@@ -64,25 +67,23 @@ namespace Quantum.Kata.SuperdenseCoding {
         // [1; 0]:    |Φ⁻⟩ = (|00⟩ - |11⟩) / sqrt(2)
         // [1; 1]:    |Ψ⁻⟩ = (|01⟩ - |10⟩) / sqrt(2)
 
+		if (message::Bit1) { // accesses the item 'Bit1' of 'message'
```
Please use spaces instead of tabs.

bettinaheim commented:
> I fixed the Notebook to refer to the data type defined in the project instead of defining its own data type (they have different namespaces), and added a few comments to help solvers figure out the syntax of dealing with data types.
> Other than that, looks good to me.

Awesome, thanks!
